File: client-unity/Assets/Scripts/Networking/Protocol.cs
namespace Ludo.Networking
{
    /// <summary>
    /// Wire protocol — must match server/src/socket/protocol.js.
    /// Keeping these as constants (not enums) avoids allocation when emitting.
    /// </summary>
    public static class ClientEvents
    {
        public const string MmPrivateCreate = "mm:private:create";
        public const string MmPrivateJoin   = "mm:private:join";
        public const string MmOnlineJoin    = "mm:online:join";
        public const string MmOnlineLeave   = "mm:online:leave";
        public const string MmBotStart      = "mm:bot:start";

        public const string GameResume      = "game:resume";
        public const string GameRoll        = "game:roll";
        public const string GameMove        = "game:move";
        public const string GameLeave       = "game:leave";

        public const string Ping            = "ping";
    }

    public static class ServerEvents
    {
        public const string MmPrivateCreated = "mm:private:created";
        public const string MmPrivateJoined  = "mm:private:joined";
        public const string MmQueued         = "mm:queued";
        public const string MmMatchFound     = "mm:match:found";

        public const string GameState        = "game:state";
        public const string GameEvent        = "game:event";
        public const string GameError        = "game:error";

        public const string Pong             = "pong";
        public const string ReconnectOk      = "reconnect:ok";
        public const string Kick             = "kick";
    }

    public static class EngineEventTypes
    {
        public const string DiceRolled    = "DICE_ROLLED";
        public const string NoMove        = "NO_MOVE";
        public const string TripleSix     = "TRIPLE_SIX";
        public const string TurnPassed    = "TURN_PASSED";
        public const string TokenOut      = "TOKEN_OUT";
        public const string TokenMoved    = "TOKEN_MOVED";
        public const string TokenKilled   = "TOKEN_KILLED";
        public const string TokenFinished = "TOKEN_FINISHED";
        public const string ExtraTurn     = "EXTRA_TURN";
        public const string SeatFinished  = "SEAT_FINISHED";
        public const string MatchFinished = "MATCH_FINISHED";
        public const string TurnTimeout   = "TURN_TIMEOUT";
    }
}

File: server/src/socket/protocol.js
'use strict';

/**
 * Wire protocol constants. Single source of truth shared with the Unity client
 * (mirrored in client-unity/Assets/Scripts/Networking/Protocol.cs).
 */

const ClientEvents = Object.freeze({
  // matchmaking
  MM_PRIVATE_CREATE: 'mm:private:create',     // {playerCount}
  MM_PRIVATE_JOIN:   'mm:private:join',       // {code}
  MM_ONLINE_JOIN:    'mm:online:join',        // {playerCount}
  MM_ONLINE_LEAVE:   'mm:online:leave',       // {playerCount}
  MM_BOT_START:      'mm:bot:start',          // {difficulty}

  // game
  GAME_RESUME:       'game:resume',           // {matchId}
  GAME_ROLL:         'game:roll',             // {actionId, matchId}
  GAME_MOVE:         'game:move',             // {actionId, matchId, tokenIndex}
  GAME_LEAVE:        'game:leave',            // {matchId}

  // misc
  PING:              'ping',                  // {t}
});

const ServerEvents = Object.freeze({
  // matchmaking
  MM_PRIVATE_CREATED: 'mm:private:created',   // {matchId, code, state}
  MM_PRIVATE_JOINED:  'mm:private:joined',    // {matchId, state}
  MM_QUEUED:          'mm:queued',            // {playerCount, position}
  MM_MATCH_FOUND:     'mm:match:found',       // {matchId, state}

  // game state
  GAME_STATE:         'game:state',           // full snapshot
  GAME_EVENT:         'game:event',           // single delta event from engine
  GAME_ERROR:         'game:error',           // {code}

  // session / connection
  PONG:               'pong',                 // {t, server}
  RECONNECT_OK:       'reconnect:ok',
  KICK:               'kick',                 // {reason}
});

// Engine event types that may appear inside `game:event`/`game:state`:
//   DICE_ROLLED, NO_MOVE, TRIPLE_SIX, TURN_PASSED, TOKEN_OUT, TOKEN_MOVED,
//   TOKEN_KILLED, TOKEN_FINISHED, EXTRA_TURN, SEAT_FINISHED, MATCH_FINISHED,
//   TURN_TIMEOUT.

module.exports = { ClientEvents, ServerEvents };
